File: calendars/src/main/java/com/rands/lightbulb/calendars/month/MonthsAdapter.java
package com.rands.lightbulb.calendars.month;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.TextView;

import com.rands.lightbulb.calendars.R;
import com.rands.lightbulb.commons.utils.ResourceUtils;
import com.rands.java.commons.date.DateUtils;

import java.util.ArrayList;

import androidx.annotation.NonNull;
import androidx.core.graphics.BlendModeColorFilterCompat;
import androidx.core.graphics.BlendModeCompat;
import androidx.recyclerview.widget.RecyclerView;

public class MonthsAdapter extends RecyclerView.Adapter<MonthsAdapter.MonthVH> {
    private ArrayList<MonthItem> items;
    private MonthCalendarView calendarView;

    public MonthsAdapter(MonthCalendarView calendarView, ArrayList<MonthItem> items) {
        this.items = items;
        this.calendarView = calendarView;
    }

    @NonNull
    @Override
    public MonthVH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        TextView v = (TextView) LayoutInflater.from(parent.getContext())
                .inflate(R.layout.calendar_month_item, parent, false);
        return new MonthVH(v);
    }

    @Override
    public void onBindViewHolder(@NonNull MonthVH holder, int position) {
        MonthItem item = items.get(position);
        holder.bindItem(item, calendarView);
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    private int findMonthPosition(int month) {
        int position = -1;
        for (int pos = 0; pos < items.size(); pos++) {
            MonthItem item = items.get(pos);
            if (item.getCurrentMonth() == month) {
                position = pos;
                break;
            }
        }
        return position;
    }

    public void selectMonth(int month) {
        int positionOfMonth = findMonthPosition(month);
        if (positionOfMonth == -1)
            return;
        clearSelection();
        items.get(positionOfMonth).setSelected(true);
        notifyItemChanged(positionOfMonth);
    }

    public void clearSelection() {
        for (int i = 0; i < items.size(); i++) {
            MonthItem item = items.get(i);
            if (!item.isSelected())
                continue;
            item.setSelected(false);
            notifyItemChanged(i);
        }
    }

    public static class MonthVH extends RecyclerView.ViewHolder {
        protected TextView monthView;

        MonthVH(TextView view) {
            super(view);
            this.monthView = (TextView) view;


        }

        public void bindItem(MonthItem item, MonthCalendarView calendarView) {
            Context context = monthView.getContext();
            monthView.setLayoutParams(new ViewGroup.LayoutParams(calendarView.getTileSize(), calendarView.getTileSize()));
            monthView.setOnClickListener(v -> calendarView.setSelectedMonth(item.currentYear, item.currentMonth));
            monthView.setEnabled(item.enabled);
            monthView.setText(item.getMonthName());
            if (item.isSelected()) {
                Drawable background = ResourceUtils.getDrawable(context, R.drawable.calendar_selected_item_background);
                background.setColorFilter(BlendModeColorFilterCompat.createBlendModeColorFilterCompat(calendarView.getBackgroundColorSelected(), BlendModeCompat.SRC_ATOP));
                monthView.setBackground(background);
            } else {
                monthView.setBackground(null);
            }
        }
    }

    public static class MonthItem {
        private final int currentYear;
        private final int currentMonth;
        private boolean selected;
        private final boolean enabled;
        private final String monthName;

        public MonthItem(int currentYear, int currentMonth, boolean selected, boolean enabled) {
            this.currentYear = currentYear;
            this.currentMonth = currentMonth;
            this.selected = selected;
            this.enabled = enabled;
            this.monthName = DateUtils.getDateString("MMM", DateUtils.date(currentYear, currentMonth));
        }

        public void setSelected(boolean selected) {
            this.selected = selected;
        }

        public int getCurrentYear() {
            return currentYear;
        }

        public int getCurrentMonth() {
            return currentMonth;
        }

        public boolean isSelected() {
            return selected;
        }

        public boolean isEnabled() {
            return enabled;
        }

        public String getMonthName() {
            return monthName;
        }
    }
}

File: lightbulb-calendars/src/main/java/com/github/rooneyandshadows/lightbulb/calendars/month/adapter/MonthsAdapter.kt
package com.github.rooneyandshadows.lightbulb.calendars.month.adapter

import android.graphics.drawable.Drawable
import android.os.Parcel
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.graphics.BlendModeColorFilterCompat
import androidx.core.graphics.BlendModeCompat
import androidx.recyclerview.widget.RecyclerView.*
import com.github.rooneyandshadows.lightbulb.calendars.month.MonthCalendarView
import com.github.rooneyandshadows.lightbulb.calendars.month.adapter.MonthsAdapter.MonthVH
import com.github.rooneyandshadows.lightbulb.calendars.R
import com.github.rooneyandshadows.lightbulb.commons.utils.ParcelUtils
import com.github.rooneyandshadows.lightbulb.commons.utils.ResourceUtils

internal class MonthsAdapter(
    private val calendarView: MonthCalendarView,
    private val items: MutableList<MonthItem>,
) : Adapter<MonthVH>() {

    @Override
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MonthVH {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.calendar_month_item, parent, false) as TextView
        return MonthVH(v)
    }

    @Override
    override fun onBindViewHolder(holder: MonthVH, position: Int) {
        val item = items[position]
        holder.bindItem(item, calendarView)
    }

    @Override
    override fun getItemCount(): Int {
        return items.size
    }

    private fun findMonthPosition(month: Int): Int {
        var position = -1
        for (pos in items.indices) {
            val item = items[pos]
            if (item.monthEntry.month == month) {
                position = pos
                break
            }
        }
        return position
    }

    fun selectMonth(month: Int) {
        val positionOfMonth = findMonthPosition(month)
        if (positionOfMonth == -1) return
        clearSelection()
        items[positionOfMonth].isSelected = true
        notifyItemChanged(positionOfMonth)
    }

    fun clearSelection() {
        for (i in items.indices) {
            val item = items[i]
            if (!item.isSelected) continue
            item.isSelected = false
            notifyItemChanged(i)
        }
    }

    class MonthVH internal constructor(private val monthView: TextView) : ViewHolder(monthView) {

        fun bindItem(item: MonthItem, calendarView: MonthCalendarView) {
            monthView.apply {
                layoutParams = ViewGroup.LayoutParams(calendarView.tileSize, calendarView.tileSize)
                setOnClickListener { calendarView.setSelectedMonth(item.monthEntry) }
                isEnabled = item.isEnabled
                text = item.monthEntry.monthName
                background = if (item.isSelected) {
                    ResourceUtils.getDrawable(context, R.drawable.calendar_selected_item_background)!!.apply {
                        colorFilter = BlendModeColorFilterCompat.createBlendModeColorFilterCompat(
                            calendarView.backgroundColorSelected,
                            BlendModeCompat.SRC_ATOP
                        )
                    }
                } else null
            }
        }
    }

    class MonthItem(
        val monthEntry: MonthEntry,
        var isSelected: Boolean,
        val isEnabled: Boolean,
    ) : Parcelable {

        constructor(parcel: Parcel) : this(
            ParcelUtils.readParcelable(parcel, MonthEntry::class.java)!!,
            ParcelUtils.readBoolean(parcel)!!,
            ParcelUtils.readBoolean(parcel)!!
        )

        override fun writeToParcel(parcel: Parcel, flags: Int) {
            ParcelUtils.apply {
                writeParcelable(parcel, monthEntry)
                writeBoolean(parcel, isSelected)
                writeBoolean(parcel, isEnabled)
            }
        }

        override fun describeContents(): Int {
            return 0
        }

        companion object CREATOR : Parcelable.Creator<MonthItem> {
            override fun createFromParcel(parcel: Parcel): MonthItem {
                return MonthItem(parcel)
            }

            override fun newArray(size: Int): Array<MonthItem?> {
                return arrayOfNulls(size)
            }
        }
    }
}